### PR TITLE
[HOPS-463] and [HOPS-506]

### PR DESCRIPTION
**Pull request by smkniazi**
[HOPS-463] The transaction title for accessing client cache is set to CONCAT.Update the name to CLIENT_CACHE.

[Hops-506] Added unit tests for STO. Added Hierarchical locking

## Make sure there is no duplicate PR for this issue

* **Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added and passed (for bug fixes / features)
- [x] HOPS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit
- [x] The commit message has the following format: [HOPS-XXX] message

* **Post a link to the associated JIRA issue**
https://hopshadoop.atlassian.net/browse/HOPS-506
https://hopshadoop.atlassian.net/browse/HOPS-463

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
bug fix

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
nope

* **Other information**:

### CODE REVIEW

**Comment by berthoug** [comment]
Does this fix all the subtree operations problems or only the subtree vs subtree?
Have you done an evaluation of the performance impact?

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/io/hops/transaction/lock/BaseINodeLock.java [hunk 1]
```diff
@@ -42,22 +39,16 @@
       allLockedInodesInTx;
   private final ResolvedINodesMap resolvedINodesMap;
   private boolean isPartitionKeyAlreaySet = false;
-
-  protected static TransactionLockTypes.INodeLockType DEFAULT_INODE_LOCK_TYPE =
-      TransactionLockTypes.INodeLockType.READ_COMMITTED;
+  private boolean enableHierarchicalLocking = false;
+  private FSNamesystem namesystem = null; //only for testing
+  private boolean isTesting = false;
```
[the remark this answers is not in the document]
Comment (smkniazi) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/io/hops/transaction/lock/BaseINodeLock.java]: nope,  i have removed it

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSConfigKeys.java [hunk 2]
```diff
@@ -162,7 +157,7 @@
   
   public static final String DFS_CLIENT_RETRIES_ON_FAILURE_KEY =
       "dfs.client.max.retries.on.failure";
-  public static final int DFS_CLIENT_RETRIES_ON_FAILURE_DEFAULT = 2;
+  public static final int DFS_CLIENT_RETRIES_ON_FAILURE_DEFAULT = 0;
       //min value is 0. Better set it >= 1
```
[the remark this answers is not in the document]
Comment (smkniazi) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSConfigKeys.java]: I changed it during testing and then forgot to revert it. it is fixed now. 
this variable needs to split. one for subtree operations and one for non subtree ops. subtree ops require lot more retries. I will do a new jira for it.

**Comment by berthoug** [comment]
Please put a link to the pull request dependencies in dal and dal-impl in your description.

**Comment by berthoug** [comment]
Jenkins test this please.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java [hunk 3]
```diff
@@ -8938,5 +9084,78 @@ public void stopMonitor() {
...
+  }
+
+  private Times saveTimes(){
+    if(isTestingSTO) {
+      delays.remove();
+      Times times = new Times(delayBeforeSTOFlag, delayAfterBuildingTree);
+      delays.set(times);
+      return times;
+    }
+    throw new UnsupportedOperationException("Operation only supported for testing");
```
Comment (berthoug) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java]: Save time is called in a lot of place without checking first if it is a test, this breaks a lot of tests.

File: hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestSubtreeConflicts.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hdfs.server.namenode;
+
+import io.hops.metadata.hdfs.TablesDef;
+import io.hops.metadata.hdfs.entity.INodeIdentifier;
+import io.hops.metadata.hdfs.entity.SubTreeOperation;
+import io.hops.metadata.ndb.mysqlserver.MySQLQueryHelper;
```
Comment (berthoug) [inline hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestSubtreeConflicts.java]: MySQLQueryHelper is in the dal-impl package, this dependency makes that it is impossible to build hops without the ndb profile, we can't do that.

**Comment by berthoug** [comment]
The following tests consistently fail both on Jenkins and on my machine, while they don't fail on the master branch:
TestDFSClientRetries.testNotYetReplicatedErrors
TestAuditLogs.testAuditDenied
TestSubtreeLock.testRetry
TestDFSPermission (timeout)
TestHDFSCLI (timeout)

**Comment by smkniazi** [comment]
Should work now. All tests passing and no dependency on DAL-IMPL layer.

**Comment by berthoug** [comment]
Jenkins test this please.

**Comment by berthoug** [comment]
Jenkins test this please.

**Comment by kouzant** [comment]
Jenkins test this please.

**Comment by kouzant** [comment]
Jenkins test this please.

**Comment by kouzant** [comment]
Jenkins test this please

**Comment by berthoug** [comment]
Jenkins test this please.

**Pull request merged**